### PR TITLE
Implemented masterchain block proofs checking

### PR DESCRIPTION
**Pull request by DezzK**

### CODE REVIEW

File: ton_client/src/proofs/mod.rs [hunk 1]
```diff
@@ -192,24 +294,18 @@ impl BlockProof {
...
-                self.id(),
-            )
-        }
-
         let (virt_block, virt_block_root) = self.virtualize_block()?;
 
         if virt_block_root.repr_hash() != self.id().root_hash() {
             bail!(
-                "proof for block {} contains a Merkle proof with incorrect root hash: expected {}, found: {} ",
+                "proof for block {} contains a Merkle proof with incorrect root hash: \
```
Comment (AlexeyVavilin) [inline ton_client/src/proofs/mod.rs:307]: Duplicating check in `virtualize_block`. Remove one.

File: ton_client/src/proofs/mod.rs [hunk 2]
```diff
@@ -305,7 +407,8 @@ impl BlockProof {
             )))?;
         let _cur_validator_set = config.config(34)?
             .ok_or_else(|| Error::invalid_data(format!(
-                "proof for key block {} contains a Merkle proof without current validators config param (34)",
+                "proof for key block {} contains a Merkle proof without current validators config \
```
Comment (AlexeyVavilin) [inline ton_client/src/proofs/mod.rs:410]: No hash check for key blocks. Need to add it

**Review by AlexeyVavilin — CHANGES_REQUESTED** [review]

File: ton_client/src/proofs/mod.rs [hunk 3]
```diff
@@ -462,10 +536,101 @@ impl BlockProof {
...
+
+    let prev_ref = &blocks[0]["prev_ref"];
+    if prev_ref.is_null() {
+        return Err(
+            Error::unable_to_resolve_zerostate_root_hash("prev_ref of the block #1 is not set")
+        );
+    }
+
+    let block_root_hash_json = &blocks[0]["id"];
+    let first_master_block_root_hash = block_root_hash_json.as_str()
```
[the remark this answers is not in the document]
Comment (DezzK) [reply ton_client/src/proofs/mod.rs:598]: No, I need block_root_hash_json value in the error message.

File: ton_client/src/proofs/mod.rs [hunk 4]
```diff
@@ -462,10 +536,101 @@ impl BlockProof {
...
+        INITIAL_TRUSTED_KEY_BLOCKS.get(&network_uid.zerostate_root_hash)
+    {
+        return Ok(&hardcoded_mc_block.trusted_key_block);
+    }
+
+    Err(Error::unable_to_resolve_trusted_key_block(&network_uid.zerostate_root_hash))
+}
+
+#[async_trait::async_trait]
+pub trait ProofHelperEngine {
```
[the remark this answers is not in the document]
Comment (DezzK) [reply ton_client/src/proofs/mod.rs:633]: In order to do dependency injection without circular reference.

File: ton_client/src/proofs/mod.rs (hunk 2 above)
Comment (DezzK) [reply ton_client/src/proofs/mod.rs:410]: It was checked before in `virtualize_block()`.

File: ton_client/src/client/client.rs [hunk 5]
```diff
@@ -165,6 +173,12 @@ pub struct ClientConfig {
     pub cache_proofs: Option<bool>,
 }
 
+impl ClientConfig {
```
[the remark this answers is not in the document]
Comment (DezzK) [reply ton_client/src/client/client.rs:176]: It is a different case. The default value has to be set to `None`, but for `cache_proofs`, we assume that `None == true`.

**Review by AlexeyVavilin — APPROVED** [review]

**Pull request merged**